extjsdk/src/test/java/io/vantiq/extjsdk/TestExtensionWebSocketClient.java

```diff
@@ -0,0 +1,248 @@
+    }
+    
+    @Test
+    public void testValidifyUrl() {
+        String url = "ws://cba.com/api/v1/wsock/websocket";
+        url = client.validifyUrl(url);
+        assert url.equals("ws://cba.com/api/v1/wsock/websocket");
+        
+        url = "http://prod.vantiq.com/api/v1/wsock/websocket";
+        url = client.validifyUrl(url);
+        assert url.equals("wss://prod.vantiq.com/api/v1/wsock/websocket");
+        
+        url = "http://prod.vantiq.com/api/v/wsock/websocket";
+        url = client.validifyUrl(url);
+        assert url.equals("wss://prod.vantiq.com/api/v/wsock/websocket/api/v1/wsock/websocket");
+        
+        url = "http://prod.vantiq.com/api/v47/wsock/websocket";
+        url = client.validifyUrl(url);
+        assert url.equals("wss://prod.vantiq.com/api/v47/wsock/websocket");
+        
+        url = "https://dev.vantiq.com";
+        url = client.validifyUrl(url);
+        assert url.equals("wss://dev.vantiq.com/api/v1/wsock/websocket");
+        
+        url = "https://dev.vantiq.com/";
+        url = client.validifyUrl(url);
+        assert url.equals("wss://dev.vantiq.com/api/v1/wsock/websocket");
+        
+        url = "dev.vantiq.com";
+        url = client.validifyUrl(url);
+        assert url.equals("wss://dev.vantiq.com/api/v1/wsock/websocket");
+    }
+    
+    @Test
+    public void testAuthenticateWithUsername() throws InterruptedException {
+        String user = "myName";
+        String pass = "p@s$w0rd";
+        
+        client.webSocketFuture.complete(true);
+        
+        client.authenticate(user, pass);
+        // do twice because the initial attempt is asynchronous, so the data will not be ready for the asserts 
+        client.authenticate(user, pass);  
+        
+        assert socket.compareData("object.username", user);
+        assert socket.compareData("object.password", pass);
+        assert socket.compareData("op", "authenticate");
+        assert socket.compareData("resourceName", "system.credentials");
+    }
+    
+    @Test
+    public void testAuthenticateWithToken() throws InterruptedException {
+        client.webSocketFuture.complete(true);
+        
+        String token = "ajeoslvkencmvkejshegwt=";
+        client.authenticate(token);
+        
+        // do twice because the initial attempt is asynchronous, so the data will not be ready for the asserts 
+        client.authenticate(token);  
+        
+        assert socket.compareData("object", token);
+        assert socket.compareData("op", "validate");
+        assert socket.compareData("resourceName", "system.credentials");
+    }
+    
+    @Test
+    public void testConnectToSource() throws InterruptedException {
+        client.authFuture.complete(true);
+        client.authSuccess.complete(null);
+        
+        client.connectToSource();
+        // do twice because the initial attempt is asynchronous, so the data will not be ready for the asserts 
+        client.connectToSource();  
```

Comment:
I feel like there should be a better way to wait for an asynchronous function to run, but I don't know what.

Reply:
Well, if it's a test, Thread.sleep() works. :-)

Otherwise, the trick is usually to have connectToSource() return it's completable, then just do a .get() on it -- to force it to run.

Reply:
Ok. I tried Thread.sleep() in one of tests, and it seems like it's a problem with that test alone, because sleeping works for the others. I'll look into it and try to add the fix in with the reconnection PR.